criu/autofs.c

```diff
@@ -62,25 +62,53 @@ int autofs_parse(struct mount_info *pm)
 {
 	long pipe_ino = AUTOFS_OPT_UNKNOWN;
 	char **opts;
-	int nr_opts, i;
+	int nr_opts, i, ret;
 
 	split(pm->options, ',', &opts, &nr_opts);
 	if (!opts)
 		return -1;
+
 	for (i = 0; i < nr_opts; i++) {
 		if (!strncmp(opts[i], "pipe_ino=", strlen("pipe_ino=")))
-			pipe_ino = atoi(opts[i] + strlen("pipe_ino="));
+			if ((errno = -xatol(opts[i] + strlen("pipe_ino="), &pipe_ino))) {
+				pr_perror("pipe_ino (%s) mount option parse failed", opts[i] + strlen("pipe_ino="));
+				ret = -1;
+				goto free;
+			}
 	}
-	for (i = 0; i < nr_opts; i++)
-		xfree(opts[i]);
-	free(opts);
+
+	/*
+	 * We must inform user about bug if pipe_ino
+	 * is greater than UINT32_MAX, because, from kernel side of view
+	 * it is possible situation, because i_ino field in struct inode
+	 * (look at include/linux/fs.h in Linux kernel) has unsigned long type.
+	 * But, at the moment, pipe_id field in pipe-data.proto has uint32 type.
+	 * 
+	 * I know, you want to ask "Why we simply not change type in image to uint64"? ;)
+	 * Let me explain this moment. We need to look at fs/inode.c in Linux kernel.
+	 * There is a get_next_ino() function, take a look on it's return type =>
+	 * currently, it has unsigned int type. Looks weird. struct inode i_ino has
+	 * unsigned long type, but get_next_ino() returns only unsigned int.
+	 * (Read big comment above get_next_ino in Linux kernel for explanation)
+	 * 
+	 * So, we don't need to do anything with our images at that moment.
+	 */
+	BUG_ON(pipe_ino > UINT32_MAX);
```

Comment:
What if pipe_ino < 0 (big enough unsigned long can overflow long)? And may be we need the same BUG_ON in parse_options() too.

Reply:
Why should it be `BUG_ON()` in the first place?
Instead of crashing CRIU, could we return error-code and handle it?

Reply:
> What if pipe_ino < 0 (big enough unsigned long can overflow long)? And may be we need the same BUG_ON in parse_options() too.

Good point, but it doesn't possible at the moment, because:
https://github.com/torvalds/linux/blob/ac309e7744bee222df6de0122facaf2d9706fa70/fs/autofs/inode.c#L93

I'm not sure, should we or not to make checks for that case. It's up to you.

Reply:
@0x7f454c46 thanks, I've replaced `BUG_ON(..)` to appropriate check